File: DocManagementBackend/Models/document.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;

namespace DocManagementBackend.Models
{
    public class Document
    {
        [Key]
        public int Id { get; set; }
        [Required]
        public int CreatedByUserId { get; set; }
        [ForeignKey("CreatedByUserId")]
        public required User CreatedBy { get; set; }
        public int? UpdatedByUserId { get; set; }
        [ForeignKey("UpdatedByUserId")]
        public User? UpdatedBy { get; set; }
        public int TypeId { get; set; }
        [ForeignKey("TypeId")]
        public DocumentType? DocumentType { get; set; }

        // New SubType relationship
        public int? SubTypeId { get; set; }
        [ForeignKey("SubTypeId")]
        public SubType? SubType { get; set; }
        public int? CurrentStatusId { get; set; }
        [ForeignKey("CurrentStatusId")]
        [JsonIgnore]
        public Status? CurrentStatus { get; set; }
        public int? CurrentStepId { get; set; }
        [ForeignKey("CurrentStepId")]
        [JsonIgnore]
        public Step? CurrentStep { get; set; }
        public int? CircuitId { get; set; }
        [ForeignKey("CircuitId")]
        public Circuit? Circuit { get; set; }
        public bool IsCircuitCompleted { get; set; } = false;
        
        public int? ResponsibilityCentreId { get; set; }
        [ForeignKey("ResponsibilityCentreId")]
        [JsonIgnore]
        public ResponsibilityCentre? ResponsibilityCentre { get; set; }
        [Required]
        public string DocumentKey { get; set; } = string.Empty;
        public string DocumentAlias { get; set; } = string.Empty;
        public string Title { get; set; } = string.Empty;
        public string? Content { get; set; }
        [Required]
        public int Status { get; set; } // 0 = Open, 1 = Validated
        public DateTime DocDate { get; set; }
        public DateTime ComptableDate { get; set; }
        public string DocumentExterne { get; set; } = string.Empty;
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
        public int LigneCouter { get; set; } = 0;
        public bool IsDeleted { get; set; } = false;
        
        // New customer/vendor relationship
        [MaxLength(50)]
        public string? CustomerOrVendor { get; set; }
        
        // Navigation properties for Customer and Vendor (will be resolved based on DocumentType.TierType)
        [ForeignKey("CustomerOrVendor")]
        public Customer? Customer { get; set; }
        
        [ForeignKey("CustomerOrVendor")]  
        public Vendor? Vendor { get; set; }
        
        // Customer/Vendor snapshot fields for historical accuracy
        [MaxLength(50)]
        public string? CustomerVendorCode { get; set; }
        
        [MaxLength(200)]
        public string? CustomerVendorName { get; set; }
        
        [MaxLength(500)]
        public string? CustomerVendorAddress { get; set; }
        
        [MaxLength(100)]
        public string? CustomerVendorCity { get; set; }
        
        [MaxLength(100)]
        public string? CustomerVendorCountry { get; set; }
        
        // ERP Integration field
        [MaxLength(100)]
        public string? ERPDocumentCode { get; set; }
        
        [JsonIgnore]
        public ICollection<Ligne> Lignes { get; set; } = new List<Ligne>();
    }

    public class DocumentType
    {
        [Key]
        public int Id { get; set; }
        public int TypeNumber { get; set; }
        public string TypeKey { get; set; } = string.Empty;
        public string TypeName { get; set; } = string.Empty;
        public string TypeAttr { get; set; } = string.Empty;
        public int DocumentCounter { get; set; } = 0;
        public int DocCounter { get; set; } = 0;
        
        // New tier type field
        public TierType TierType { get; set; } = TierType.None;
        
        [JsonIgnore]
        public ICollection<Document> Documents { get; set; } = new List<Document>();
        [JsonIgnore]
        public ICollection<Circuit> Circuits { get; set; } = new List<Circuit>();
    }
    public class SubType
    {
        [Key]
        public int Id { get; set; }

        public string SubTypeKey { get; set; } = string.Empty;

        [Required]
        public string Name { get; set; } = string.Empty;

        public string Description { get; set; } = string.Empty;

        [Required]
        public DateTime StartDate { get; set; }

        [Required]
        public DateTime EndDate { get; set; }

        [Required]
        public int DocumentTypeId { get; set; }

        [ForeignKey("DocumentTypeId")]
        [JsonIgnore]
        public DocumentType? DocumentType { get; set; }

        public bool IsActive { get; set; } = true;

        [JsonIgnore]
        public ICollection<Document> Documents { get; set; } = new List<Document>();
    }
    public class TypeCounter
    {
        public int Id { get; set; }
        public int Counter { get; set; }
        public int circuitCounter { get; set; }
    }
    public class DocumentCircuitHistory
    {
        [Key]
        public int Id { get; set; }
        [Required]
        public int DocumentId { get; set; }
        [ForeignKey("DocumentId")]
        [JsonIgnore]
        public Document? Document { get; set; }
        public int? StepId { get; set; }
        [ForeignKey("StepId")]
        [JsonIgnore]
        public Step? Step { get; set; }
        public int? ActionId { get; set; }
        [ForeignKey("ActionId")]
        [JsonIgnore]
        public Action? Action { get; set; }
        public int? StatusId { get; set; }
        [ForeignKey("StatusId")]
        [JsonIgnore]
        public Status? Status { get; set; }
        [Required]
        public int ProcessedByUserId { get; set; }
        [ForeignKey("ProcessedByUserId")]
        [JsonIgnore]
        public User? ProcessedBy { get; set; }
        public DateTime ProcessedAt { get; set; } = DateTime.UtcNow;
        public string Comments { get; set; } = string.Empty;
        public bool IsApproved { get; set; } = true;
    }
    
    public class DocumentStepHistory
    {
        [Key]
        public int Id { get; set; }
        [Required]
        public int DocumentId { get; set; }
        [ForeignKey("DocumentId")]
        [JsonIgnore]
        public Document? Document { get; set; }
        [Required]
        public int StepId { get; set; }
        [ForeignKey("StepId")]
        [JsonIgnore]
        public Step? Step { get; set; }
        [Required]
        public int UserId { get; set; }
        [ForeignKey("UserId")]
        [JsonIgnore]
        public User? User { get; set; }
        public DateTime TransitionDate { get; set; } = DateTime.UtcNow;
        public string Comments { get; set; } = string.Empty;
    }
}

File: DocManagementFrontend/src/models/document.ts
import { ResponsibilityCentreSimple } from './responsibilityCentre';
import { 
  LignesElementType, 
  Item, 
  UniteCode, 
  GeneralAccounts 
} from './lineElements';
import { SubType } from './subtype';
import { LocationDto } from './location';
import './circuit.d.ts';

export enum TierType {
  None = 0,
  Customer = 1,
  Vendor = 2
}

export interface Document {
  id: number;
  title: string;
  documentKey: string;
  content: string;
  status: number;
  documentAlias: string;
  documentExterne?: string;
  createdAt: string;
  updatedAt: string;
  typeId: number;
  subTypeId?: number;
  docDate: string;
  comptableDate: string;
  documentType: DocumentType;
  subType?: SubType;
  circuitId?: number;
  circuit?: Circuit;
  currentCircuitDetailId?: number;
  currentCircuitDetail?: CircuitDetail;
  createdByUserId: number;
  createdBy: DocumentUser;
  updatedByUserId?: number;
  updatedBy?: DocumentUser;
  lignesCount?: number;
  sousLignesCount?: number;
  lignes?: Ligne[];
  responsibilityCentreId?: number;
  responsibilityCentre?: ResponsibilityCentreSimple;
  
  // Customer/Vendor snapshot data
  customerVendorCode?: string;
  customerVendorName?: string;
  customerVendorAddress?: string;
  customerVendorCity?: string;
  customerVendorCountry?: string;
  
  // ERP Archive status
  erpDocumentCode?: string;
  isArchived?: boolean;
}

export interface DocumentType {
  id?: number;
  typeNumber?: number;
  typeName: string;
  typeKey?: string;
  typeAttr?: string;
  tierType?: TierType;
  documentCounter?: number;
  createdAt?: string;
  updatedAt?: string;
}

export interface DocumentUser {
  id: number;
  username: string;
  firstName: string;
  lastName: string;
  role: string;
  email: string;
}

export interface CreateDocumentRequest {
  title: string;
  content: string;
  documentAlias?: string;
  documentExterne?: string;
  typeId: number;
  subTypeId?: number | null;
  docDate?: string;
  circuitId?: number;
  responsibilityCentreId?: number;
  
  // Customer/Vendor information
  customerVendorCode?: string;
  customerVendorName?: string;
  customerVendorAddress?: string;
  customerVendorCity?: string;
  customerVendorCountry?: string;
}

export interface UpdateDocumentRequest {
  title?: string;
  content?: string;
  documentAlias?: string;
  documentExterne?: string;
  typeId?: number;
  docDate?: string;
  comptableDate?: string;
  circuitId?: number;
  
  // Customer/Vendor information
  customerVendorCode?: string;
  customerVendorName?: string;
  customerVendorAddress?: string;
  customerVendorCity?: string;
  customerVendorCountry?: string;
}

export interface Ligne {
  id: number;
  documentId: number;
  ligneKey: string;
  title: string;
  article: string;
  
  // Legacy field
  prix: number;
  
  sousLignesCount: number;
  
  // Type information - FIXED field names to match backend
  lignesElementTypeId?: number;
  lignesElementType?: LignesElementType;
  
  // Element references
  itemCode?: string;
  item?: Item;
  generalAccountsCode?: string;
  generalAccounts?: GeneralAccounts;
  
  // Location reference (only for Item types)
  locationCode?: string;
  location?: LocationDto;
  
  // Unit of measure reference (only for Item types)
  unitCode?: string;
  unit?: UniteCode;
  
  // Pricing fields
  quantity: number;
  priceHT: number;
  discountPercentage: number;
  discountAmount?: number;
  vatPercentage: number;
  
  // Calculated fields
  amountHT: number;
  amountVAT: number;
  amountTTC: number;
  
  // ERP Integration field
  erpLineCode?: string;
  
  createdAt: string;
  updatedAt: string;
  document?: Document;
  sousLignes?: SousLigne[];
}

export interface CreateLigneRequest {
  documentId: number;
  ligneKey?: string;
  title: string;
  article: string;
  
  // Type reference - backend determines item/account automatically based on this
  lignesElementTypeId?: number;
  
  // Selected element code (Item.Code or GeneralAccounts.Code)
  selectedElementCode?: string;
  
  // Location code (only for Item types)
  locationCode?: string;
  
  // Unit of measure code (only for Item types)
  unitCode?: string;
  
  // Pricing fields (calculated by frontend)
  quantity: number;
  priceHT: number; // Unit price after conversion (adjusted price)
  discountPercentage: number;
  discountAmount: number; // Calculated discount amount
  vatPercentage: number;
  
  // Original unit price (before conversion) - for reference
  originalPriceHT: number;
}

export interface UpdateLigneRequest {
  ligneKey?: string;
  title?: string;
  article?: string;
  
  // Type reference - backend determines item/account automatically based on this
  lignesElementTypeId?: number;
  
  // Selected element code (Item.Code or GeneralAccounts.Code)
  selectedElementCode?: string;
  
  // Location code (only for Item types)
  locationCode?: string;
  
  // Unit of measure code (only for Item types)
  unitCode?: string;
  
  // Pricing fields (calculated by frontend)
  quantity?: number;
  priceHT?: number; // Unit price after conversion (adjusted price)
  discountPercentage?: number;
  discountAmount?: number; // Calculated discount amount
  vatPercentage?: number;
  
  // Original unit price (before conversion) - for reference
  originalPriceHT?: number;
}

export interface SousLigne {
  id: number;
  ligneId: number;
  sousLigneKey?: string;
  title: string;
  attribute: string;
  createdAt?: string;
  updatedAt?: string;
  ligne?: Ligne;
}

export interface CreateSousLigneRequest {
  ligneId: number;
  title: string;
  attribute: string;
}

export interface UpdateSousLigneRequest {
  title?: string;
  attribute?: string;
}
